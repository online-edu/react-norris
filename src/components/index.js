import FavButton from './fav-button';
import Footer from './footer';
import Header from './header';
import List from './list';
import Spinner from './spinner';
import Switch from './switch';
import NotFound from './not-found';

export {
  Spinner,
  Switch,
  List,
  FavButton,
  Header,
  Footer,
  NotFound,
};
