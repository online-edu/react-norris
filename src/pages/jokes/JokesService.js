import { api } from '../../utils/config';
import request from '../../utils/request';

/**
 * Load jokes from server
 *
 * @returns {Promise}
 */
const loadJokes = (count = 10) => request.get(`${api}/random/${count}`).then(({ value: jokes }) => Promise.resolve(jokes));

/**
 * Retrieve all the jokes from `localStroage`
 *
 * @returns {Object[]}
 */
const getJokesFromStorage = () => {
  const jokes = localStorage.getItem('jokes');
  return (jokes && JSON.parse(jokes)) || [];
};

/**
 * Load jokes from localStorage
 *
 * @returns {Object[]}
 */
const loadFavoriteJokes = () => getJokesFromStorage();

/**
 * Find a joke by id from `localStroage`
 *
 * @returns {Object}
 */
const findJokeById = (id) => {
  const jokes = getJokesFromStorage();
  const found = jokes.findIndex(joke => joke.id === id);
  return { jokes, found };
};

/**
 * Add joke to localStorage to maintain favorite list
 *
 * @param {Object} joke - joke to add
 * @returns {Object[]}
 */
const addFavoriteJoke = (joke) => {
  const { jokes, found } = findJokeById(joke.id);
  if (found < 0 && jokes.length <= 10) jokes.push(joke);
  localStorage.setItem('jokes', JSON.stringify(jokes));
  return jokes;
};

/**
 * Remove joke from localStorage to maintain favorite list
 *
 * @param {number} id - id to find joke and remove
 * @returns {Object[]}
 */
const removeFavoriteJoke = (id) => {
  const { jokes, found: i } = findJokeById(id);
  if (i >= 0) jokes.splice(i, 1);
  localStorage.setItem('jokes', JSON.stringify(jokes));
  return jokes;
};

/**
 * On favorite click either add or remove joke from localStorage
 *
 * @param {Object} joke - joke to add/remove
 * @param {boolean} mode - add if true otherwise remove
 * @returns {Object[]}
 */
const onFavorite = (joke, mode) => (mode ? addFavoriteJoke(joke) : removeFavoriteJoke(joke.id));

export { loadJokes, onFavorite, loadFavoriteJokes };
