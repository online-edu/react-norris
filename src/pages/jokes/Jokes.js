import React, { Component } from 'react';
import { List, Header } from '../../components';
import auth from '../../utils/auth';
import { loadJokes, onFavorite, loadFavoriteJokes } from './JokesService';
/**
 * Jokes component.
 */
class Jokes extends Component {
  /**
   * Intilalizes the state and binds all methods.
   *
   * @param {Object} props - access 'props'
   */
  constructor(props) {
    super(props);
    this.state = {
      jokes: [],
      favoriteJokes: [],
      loading: true,
      loadingFavorite: false,
      error: false,
      errMsg: '',
    };
    this.loadJokes = this.loadJokes.bind(this);
    this.onFavoriteClick = this.onFavoriteClick.bind(this);
    this.onSwitchToggle = this.onSwitchToggle.bind(this);
    this.timer = {};
  }

  /**
   *  Fetch jokes from API and localStorage once component is mounted.
   */
  componentDidMount() {
    this.loadJokes();
    this.setState({ favoriteJokes: loadFavoriteJokes() });
  }

  /**
   * On favorite click of particular joke update the lists.
   *
   * @param {Object} joke - joke
   * @param {boolean} fav - fav or un-fav
   */
  onFavoriteClick({ item: joke, fav }) {
    const jokes = onFavorite(joke, fav);
    this.setState({ favoriteJokes: jokes });
  }

  /**
   * On switch toggle generate random jokes.
   *
   * @param {Event} e - an input event
   */
  onSwitchToggle({ target }) {
    const { checked } = target;
    if (checked) {
      this.setState({ loadingFavorite: true });
      this.timer = setInterval(() => this.ticker(), 5000);
    } else this.clearTimer();
  }

  /**
   * On every tick fetch a random joke and update list.
   */
  ticker() {
    const { favoriteJokes } = this.state;
    if (favoriteJokes.length < 10) {
      loadJokes(1)
        .then(([joke]) => this.onFavoriteClick({ item: joke, fav: true }))
        .catch(err => this.handleError(err));
    } else this.clearTimer();
  }

  /**
   * Clear the interval.
   */
  clearTimer() {
    clearInterval(this.timer);
    this.setState({ loadingFavorite: false });
  }

  /**
   * Load jokes once component is mounted.
   */
  loadJokes() {
    this.setState({ loading: true });
    loadJokes()
      .then(jokes => this.setState({ jokes, loading: false }))
      .catch(err => this.handleError(err));
  }

  /**
   * Handles common errors for n/w requests.
   *
   * @param {string} err - Error message
   */
  handleError(err) {
    this.setState({ loading: false, error: true, errMsg: err.toString() });
  }

  /**
   * Render method for component
   */
  render() {
    const {
      jokes,
      favoriteJokes,
      loading,
      loadingFavorite,
      error,
      errMsg,
    } = this.state;

    return (
      <div>
        <Header user={auth.getUser()} />
        <section className="container-fluid px-4 pt-4 mt-2">
          <div className="row">
            <div className="col-md-6 col-sm-12 mb-4">
              {error && (
                <div className="alert alert-danger" role="alert">
                  {errMsg}
                </div>
              )}
              <div className="card">
                <List
                  title={`Jokes (${jokes.length})`}
                  btnCaption="More fun"
                  loader={loading}
                  items={jokes}
                  onHeaderAction={this.loadJokes}
                  onFavoriteClick={this.onFavoriteClick}
                />
              </div>
            </div>
            <div className="col-md-6 col-sm-12 mb-4">
              <div className="card">
                <List
                  favorite
                  title="Favorite Jokes"
                  items={favoriteJokes}
                  loader={loadingFavorite}
                  onFavoriteClick={this.onFavoriteClick}
                  switchCaption="Random"
                  onSwitchToggle={this.onSwitchToggle}
                />
              </div>
            </div>
          </div>
        </section>
      </div>
    );
  }
}

export default Jokes;
