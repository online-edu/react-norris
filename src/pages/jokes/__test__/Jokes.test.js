import React from 'react';
import { shallow } from 'enzyme';
import Jokes from '../Jokes';
import { List } from '../../../components';
import { loadJokes } from '../JokesService';

describe('<Jokes />', () => {
  const joke = {
    id: 480,
    joke: 'The class object inherits from Chuck Norris',
    categories: ['nerdy'],
  };
  jest.setTimeout(30000);

  it('renders correctly', () => {
    const component = shallow(<Jokes />);
    expect(component).toMatchSnapshot();
  });

  it('renders two <List /> component', () => {
    const wrapper = shallow(<Jokes />);
    expect(wrapper.find(List)).toHaveLength(2);
  });

  it('calls componentDidMount', () => {
    jest.spyOn(Jokes.prototype, 'componentDidMount');
    const wrapper = shallow(<Jokes />);
    wrapper.update();
    expect(Jokes.prototype.componentDidMount).toBeCalledTimes(1);
  });

  it('componentDidMount should fetch, and put jokes in state', () => {
    const wrapper = shallow(<Jokes />);
    return loadJokes()
      .then((jokes) => {
        wrapper.setState({ jokes });
        expect(wrapper.state('jokes')[0]).toContain('joke');
      })
      .catch((err) => {
        wrapper.setState({ loading: false });
        expect(wrapper.state('loading')).toBe(false);
        expect(err).toBeDefined();
      });
  });

  it('makes a joke favorite on fav click', () => {
    const wrapper = shallow(<Jokes />);
    expect(wrapper.state('favoriteJokes').length).toBeLessThan(1);
    wrapper.instance().onFavoriteClick({ item: joke, fav: true });
    expect(wrapper.state('favoriteJokes').length).toBeGreaterThan(0);
  });

  it('generates a joke/5 seconds radomly', () => {
    const wrapper = shallow(<Jokes />);
    const e = {
      target: {
        checked: true,
      },
    };
    expect(wrapper.state('favoriteJokes').length).toBeLessThanOrEqual(1);
    wrapper.instance().onSwitchToggle(e);
    expect(wrapper.state('loadingFavorite')).toBe(true);
    setTimeout(() => {
      expect(wrapper.state('favoriteJokes').length).toBeGreaterThanOrEqual(1);
    }, 10000);
  });

  it('fetches a joke on tick', () => {
    const wrapper = shallow(<Jokes />);
    expect(wrapper.state('favoriteJokes').length).toBeLessThanOrEqual(1);
    wrapper.instance().ticker();
    setTimeout(() => {
      expect(wrapper.state('favoriteJokes').length).toBeGreaterThanOrEqual(1);
    }, 1000);
  });

  it('fetch a joke and set in list', () => {
    const wrapper = shallow(<Jokes />);
    return loadJokes(1).then((tempJoke) => {
      wrapper.instance().onFavoriteClick({ item: tempJoke, fav: true });
      expect(wrapper.state('favoriteJokes').length).toBeGreaterThan(0);
    });
  });

  it('handles error', () => {
    const wrapper = shallow(<Jokes />);
    const testPromise = new Promise((res, reject) => {
      setTimeout(() => reject(Error({ err: 'Promise test' })), 1000);
    });
    return testPromise.catch((err) => {
      wrapper.instance().handleError(err);
      expect(wrapper.state('error')).toBe(true);
    });
  });

  it('stops time on switch off', () => {
    const wrapper = shallow(<Jokes />);
    const e = {
      target: {
        checked: true,
      },
    };
    wrapper.instance().onSwitchToggle(e);
    expect(wrapper.state('loadingFavorite')).toBe(true);
    wrapper.instance().clearTimer();
    expect(wrapper.state('loadingFavorite')).toBe(false);
  });
});
