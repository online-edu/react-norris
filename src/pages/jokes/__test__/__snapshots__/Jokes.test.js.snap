// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Jokes /> renders correctly 1`] = `ShallowWrapper {}`;
